# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/shell/js.cpp
# Commit: 9d31a829691c
# Full Hash: 9d31a829691ccf798375446e7ef04109e2662175
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2023-06-07 03:30:33
# Regressor Bug: 1835876
# File Overlap Count: 1
# Description:
#   Bug 1835876 part 2 - Disable code write protection in content processes. r=nbp
#   
#   We've worked with the security and performance teams to re-evaluate the W^X policy
#   we have in place to mark JIT code memory pages as either writable or executable
#   (but not both).
# ==============================================================================

diff -r 1ce5ad17da6b -r 9d31a829691c js/src/shell/js.cpp
--- a/js/src/shell/js.cpp	Tue Jun 06 12:16:46 2023 +0000
+++ b/js/src/shell/js.cpp	Tue Jun 06 12:16:46 2023 +0000
@@ -11411,6 +11411,11 @@
       !op.addStringOption('\0', "spectre-mitigations", "on/off",
                           "Whether Spectre mitigations are enabled (default: "
                           "off, on to enable)") ||
+      !op.addStringOption('\0', "write-protect-code", "on/off",
+                          "Whether the W^X policy is enforced to mark JIT code "
+                          "pages as either writable or executable but never "
+                          "both at the same time (default: on, off to "
+                          "disable)") ||
       !op.addStringOption('\0', "cache-ir-stubs", "on/off/call",
                           "Use CacheIR stubs (default: on, off to disable, "
                           "call to enable work-in-progress call ICs)") ||
@@ -12130,6 +12135,16 @@
     }
   }
 
+  if (const char* str = op.getStringOption("write-protect-code")) {
+    if (strcmp(str, "on") == 0) {
+      jit::JitOptions.writeProtectCode = true;
+    } else if (strcmp(str, "off") == 0) {
+      jit::JitOptions.writeProtectCode = false;
+    } else {
+      return OptionFailure("write-protect-code", str);
+    }
+  }
+
   if (const char* str = op.getStringOption("ion-scalar-replacement")) {
     if (strcmp(str, "on") == 0) {
       jit::JitOptions.disableScalarReplacement = false;